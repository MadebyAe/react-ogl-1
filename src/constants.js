import * as OGL from 'ogl'

/**
 * OGL elements which must accept `gl` via constructor args.
 */
export const GL_ELEMENTS = [OGL.Camera, OGL.Geometry, OGL.Mesh, OGL.Program]

/**
 * React's internal props.
 */
export const RESERVED_PROPS = ['children', 'key', 'ref', '__self', '__source']

/**
 * React rendering modes (defaults to blocking).
 */
export const RENDER_MODES = {
  legacy: 0,
  blocking: 1,
  concurrent: 2,
}
