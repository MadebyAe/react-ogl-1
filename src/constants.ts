// @ts-ignore
import * as OGL from 'ogl'

/**
 * OGL elements which must accept `gl` via constructor args.
 */
export const GL_ELEMENTS = [
  OGL.Camera,
  OGL.Geometry,
  OGL.Mesh,
  OGL.Program,
  OGL.Texture,
  OGL.Polyline,
  OGL.NormalProgram,
] as const

/**
 * React's internal props.
 */
export const RESERVED_PROPS = ['children', 'key', 'ref', '__self', '__source'] as const

/**
 * React rendering modes (defaults to blocking).
 */
export const RENDER_MODES = {
  legacy: 0,
  blocking: 1,
  concurrent: 2,
} as const
