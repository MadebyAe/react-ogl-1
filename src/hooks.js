import { useLayoutEffect, createContext, useContext, useRef } from 'react'

/**
 * Internal OGL context.
 */
export const OGLContext = createContext(null)

/**
 * Accesses internal OGL state.
 */
export const useOGL = () => {
  const state = useContext(OGLContext)
  // We can only access context from within the scope of a context provider.
  // If used outside, we throw an error instead of returning null for DX.
  if (!state) throw 'Hooks must used inside a canvas or OGLContext provider!'
  return state
}

/**
 * Subscribe an element into a shared render loop.
 */
export const useFrame = (callback, renderPriority = 0) => {
  const state = useOGL()
  // Store frame callback in a ref so we can pass a mutable reference.
  // This allows the callback to dynamically update without blocking
  // the render loop.
  const ref = useRef(callback)
  useLayoutEffect(() => void (ref.current = callback), [callback])
  // Subscribe on mount and unsubscribe on unmount
  useLayoutEffect(() => {
    state.subscribe(ref, renderPriority)
    return () => state.unsubscribe(ref, renderPriority)
  }, [state, renderPriority])
}
