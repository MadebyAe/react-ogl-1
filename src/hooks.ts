import { useLayoutEffect, createContext, useContext, useRef } from 'react'
import { RootState, Subscription } from './types'

/**
 * Internal OGL context.
 */
export const OGLContext = createContext(null)

/**
 * Accesses internal OGL state.
 */
export const useOGL = () => {
  const state: RootState = useContext(OGLContext)
  // We can only access context from within the scope of a context provider.
  // If used outside, we throw an error instead of returning null for DX.
  if (!state) throw 'Hooks must used inside a canvas or OGLContext provider!'
  return state
}

/**
 * Subscribe an element into a shared render loop.
 */
export const useFrame = (callback: Subscription, renderPriority = 0) => {
  const state = useOGL()
  // Store frame callback in a ref so we can pass a mutable reference.
  // This allows the callback to dynamically update without blocking
  // the render loop.
  const ref = useRef(callback)
  useLayoutEffect(() => void (ref.current = callback), [callback])
  // Subscribe on mount and unsubscribe on unmount
  useLayoutEffect(() => {
    state.subscribe(ref, renderPriority)
    return () => void state.unsubscribe(ref, renderPriority)
  }, [state, renderPriority])
}
