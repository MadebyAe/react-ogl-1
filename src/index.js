export * from './constants'
export * from './renderer'
export * from './utils'
