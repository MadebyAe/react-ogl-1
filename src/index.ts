export * from './constants'
export * from './hooks'
export * from './reconciler'
export * from './renderer'
export * from './types'
export * from './utils'
