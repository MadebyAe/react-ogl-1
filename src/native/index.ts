export * from './Canvas'
export * from './events'
export * from './hooks'
export * from '../'
