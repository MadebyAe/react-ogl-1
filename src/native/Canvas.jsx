import { forwardRef, useRef, useState, useEffect, Suspense } from 'react'
import { PixelRatio, View } from 'react-native'
import { GLView } from 'expo-gl'
import { useIsomorphicLayoutEffect } from '../shared/hooks'
import { createInternals } from '../shared/utils'
import { ErrorBoundary, Block } from '../shared/components'
import { events } from './events'
import { filterKeys } from '../utils'
import { RESERVED_PROPS } from '../constants'

/**
 * A list of custom Canvas props.
 */
export const CANVAS_PROPS = [
  'renderer',
  'camera',
  'orthographic',
  'frameloop',
  'events',
  'onCreated',
]

/**
 * A resizeable canvas whose children are declarative OGL elements.
 */
export const Canvas = forwardRef(({ children, style, ...rest }, forwardedRef) => {
  const internalProps = filterKeys(rest, false, ...CANVAS_PROPS)
  const viewProps = filterKeys(rest, true, ...CANVAS_PROPS, ...RESERVED_PROPS)
  const internalState = useRef()
  const [{ width, height }, setSize] = useState({ width: 0, height: 0 })
  const [context, setContext] = useState()
  const [bind, setBind] = useState()
  const [block, setBlock] = useState(false)
  const [error, setError] = useState(false)

  // Suspend this component if block is a promise (2nd run)
  if (block) throw block
  // Throw exception outwards if anything within Canvas throws
  if (error) throw error

  // Execute JSX in the reconciler as a layout-effect
  useIsomorphicLayoutEffect(() => {
    if (width > 0 && height > 0 && context) {
      // If first run, create default state
      if (!internalState.current) {
        internalState.current = createInternals(view.current, {
          events,
          ...internalProps,
          renderer: { ...internalProps?.renderer, context },
        })
      }

      const state = internalState.current

      // Set dpr, handle resize
      state.renderer.dpr = PixelRatio.get()
      state.renderer.setSize(width, height)

      // Update projection
      const projection = internalProps.orthographic ? 'orthographic' : 'perspective'
      state.camera[projection]({ aspect: width / height })

      // Render to screen
      state.root.render(
        <ErrorBoundary set={setError}>
          <Suspense fallback={<Block set={setBlock} />}>{children}</Suspense>
        </ErrorBoundary>,
      )
      setBind(state.events.connected.getEventHandlers())
    }
  }, [width, height, context, internalProps, children])

  // Cleanup on unmount
  useEffect(() => {
    const state = internalState.current
    return () => state?.root.unmount()
  }, [])

  return (
    <View
      {...viewProps}
      ref={forwardedRef}
      onLayout={(event) => {
        const { width, height } = event.nativeEvent.layout
        setSize({ width, height })
      }}
      style={{ flex: 1, ...style }}
      {...bind}
    >
      {width > 0 && (
        <GLView onContextCreate={setContext} style={StyleSheet.absoluteFill} />
      )}
    </View>
  )
})
