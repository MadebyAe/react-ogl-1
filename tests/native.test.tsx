import * as React from 'react'
import { View } from 'react-native'
import { render, RenderAPI } from '@testing-library/react-native'
import { reconciler, Canvas } from '../src/native'

describe('Canvas', () => {
  it('should correctly mount', async () => {
    let renderer: RenderAPI

    await reconciler.act(async () => {
      renderer = render(
        <Canvas>
          <transform />
        </Canvas>,
      )
    })

    expect(renderer.container).toMatchSnapshot()
  })

  it('should forward ref', async () => {
    const ref = React.createRef<View>()

    await reconciler.act(async () => {
      render(
        <Canvas ref={ref}>
          <transform />
        </Canvas>,
      )
    })

    expect(ref.current).toBeDefined()
  })

  it('should correctly unmount', async () => {
    let renderer: RenderAPI

    await reconciler.act(async () => {
      renderer = render(
        <Canvas>
          <transform />
        </Canvas>,
      )
    })

    expect(() => renderer.unmount()).not.toThrow()
  })
})
