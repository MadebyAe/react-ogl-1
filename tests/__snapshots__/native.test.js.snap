// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Canvas should correctly mount 1`] = `
ReactTestInstance {
  "_fiber": FiberNode {
    "_debugHookTypes": Array [
      "useRef",
      "useState",
      "useState",
      "useState",
      "useState",
      "useState",
      "useLayoutEffect",
      "useEffect",
    ],
    "_debugID": 3,
    "_debugNeedsRemount": false,
    "_debugOwner": null,
    "_debugSource": null,
    "actualDuration": 0,
    "actualStartTime": -1,
    "alternate": null,
    "child": FiberNode {
      "_debugHookTypes": null,
      "_debugID": 4,
      "_debugNeedsRemount": false,
      "_debugOwner": [Circular],
      "_debugSource": null,
      "actualDuration": 0,
      "actualStartTime": -1,
      "alternate": null,
      "child": null,
      "childLanes": 0,
      "dependencies": null,
      "elementType": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
      "firstEffect": null,
      "flags": 1,
      "index": 0,
      "key": null,
      "lanes": 0,
      "lastEffect": null,
      "memoizedProps": Object {
        "children": false,
        "onLayout": [Function],
        "style": Object {
          "flex": 1,
        },
      },
      "memoizedState": null,
      "mode": 0,
      "nextEffect": null,
      "pendingProps": Object {
        "children": false,
        "onLayout": [Function],
        "style": Object {
          "flex": 1,
        },
      },
      "ref": null,
      "return": [Circular],
      "selfBaseDuration": 0,
      "sibling": null,
      "stateNode": null,
      "tag": 11,
      "treeBaseDuration": 0,
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
      "updateQueue": null,
    },
    "childLanes": 0,
    "dependencies": null,
    "elementType": Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    },
    "firstEffect": null,
    "flags": 517,
    "index": 0,
    "key": null,
    "lanes": 0,
    "lastEffect": null,
    "memoizedProps": Object {
      "children": <transform />,
    },
    "memoizedState": Object {
      "baseQueue": null,
      "baseState": null,
      "memoizedState": Object {
        "current": undefined,
      },
      "next": Object {
        "baseQueue": null,
        "baseState": Object {
          "height": 0,
          "width": 0,
        },
        "memoizedState": Object {
          "height": 0,
          "width": 0,
        },
        "next": Object {
          "baseQueue": null,
          "baseState": undefined,
          "memoizedState": undefined,
          "next": Object {
            "baseQueue": null,
            "baseState": undefined,
            "memoizedState": undefined,
            "next": Object {
              "baseQueue": null,
              "baseState": false,
              "memoizedState": false,
              "next": Object {
                "baseQueue": null,
                "baseState": false,
                "memoizedState": false,
                "next": Object {
                  "baseQueue": null,
                  "baseState": null,
                  "memoizedState": Object {
                    "create": [Function],
                    "deps": Array [
                      0,
                      0,
                      undefined,
                      Object {},
                      <transform />,
                    ],
                    "destroy": undefined,
                    "next": Object {
                      "create": [Function],
                      "deps": Array [],
                      "destroy": [Function],
                      "next": [Circular],
                      "tag": 5,
                    },
                    "tag": 3,
                  },
                  "next": Object {
                    "baseQueue": null,
                    "baseState": null,
                    "memoizedState": Object {
                      "create": [Function],
                      "deps": Array [],
                      "destroy": [Function],
                      "next": Object {
                        "create": [Function],
                        "deps": Array [
                          0,
                          0,
                          undefined,
                          Object {},
                          <transform />,
                        ],
                        "destroy": undefined,
                        "next": [Circular],
                        "tag": 3,
                      },
                      "tag": 5,
                    },
                    "next": null,
                    "queue": null,
                  },
                  "queue": null,
                },
                "queue": Object {
                  "dispatch": [Function],
                  "lastRenderedReducer": [Function],
                  "lastRenderedState": false,
                  "pending": null,
                },
              },
              "queue": Object {
                "dispatch": [Function],
                "lastRenderedReducer": [Function],
                "lastRenderedState": false,
                "pending": null,
              },
            },
            "queue": Object {
              "dispatch": [Function],
              "lastRenderedReducer": [Function],
              "lastRenderedState": undefined,
              "pending": null,
            },
          },
          "queue": Object {
            "dispatch": [Function],
            "lastRenderedReducer": [Function],
            "lastRenderedState": undefined,
            "pending": null,
          },
        },
        "queue": Object {
          "dispatch": [Function],
          "lastRenderedReducer": [Function],
          "lastRenderedState": Object {
            "height": 0,
            "width": 0,
          },
          "pending": null,
        },
      },
      "queue": null,
    },
    "mode": 0,
    "nextEffect": null,
    "pendingProps": Object {
      "children": <transform />,
    },
    "ref": null,
    "return": FiberNode {
      "_debugHookTypes": null,
      "_debugID": 1,
      "_debugNeedsRemount": false,
      "_debugOwner": null,
      "_debugSource": null,
      "actualDuration": 0,
      "actualStartTime": -1,
      "alternate": FiberNode {
        "_debugHookTypes": null,
        "_debugID": 1,
        "_debugNeedsRemount": false,
        "_debugOwner": null,
        "_debugSource": null,
        "actualDuration": 0,
        "actualStartTime": -1,
        "alternate": [Circular],
        "child": null,
        "childLanes": 0,
        "dependencies": null,
        "elementType": null,
        "firstEffect": null,
        "flags": 0,
        "index": 0,
        "key": null,
        "lanes": 1,
        "lastEffect": null,
        "memoizedProps": null,
        "memoizedState": null,
        "mode": 0,
        "nextEffect": null,
        "pendingProps": null,
        "ref": null,
        "return": null,
        "selfBaseDuration": 0,
        "sibling": null,
        "stateNode": FiberRootNode {
          "_debugRootType": "createLegacyRoot()",
          "callbackNode": null,
          "callbackPriority": 15,
          "containerInfo": Object {
            "children": Array [],
            "createNodeMock": [Function],
            "tag": "CONTAINER",
          },
          "context": Object {},
          "current": [Circular],
          "entangledLanes": 0,
          "entanglements": Array [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
          ],
          "eventTimes": Array [
            -1,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
          ],
          "expirationTimes": Array [
            -1,
            -1,
            -1,
            -1,
            -1,
            -1,
            -1,
            -1,
            -1,
            -1,
            -1,
            -1,
            -1,
            -1,
            -1,
            -1,
            -1,
            -1,
            -1,
            -1,
            -1,
            -1,
            -1,
            -1,
            -1,
            -1,
            -1,
            -1,
            -1,
            -1,
            -1,
          ],
          "expiredLanes": 0,
          "finishedLanes": 0,
          "finishedWork": null,
          "hydrate": false,
          "interactionThreadID": 1,
          "memoizedInteractions": Set {},
          "mutableReadLanes": 0,
          "pendingChildren": null,
          "pendingContext": null,
          "pendingInteractionMap": Map {},
          "pendingLanes": 0,
          "pingCache": null,
          "pingedLanes": 0,
          "suspendedLanes": 0,
          "tag": 0,
          "timeoutHandle": -1,
        },
        "tag": 3,
        "treeBaseDuration": 0,
        "type": null,
        "updateQueue": Object {
          "baseState": null,
          "effects": null,
          "firstBaseUpdate": Object {
            "callback": null,
            "eventTime": 0,
            "lane": 1,
            "next": null,
            "payload": Object {
              "element": <ForwardRef>
                <transform />
              </ForwardRef>,
            },
            "tag": 0,
          },
          "lastBaseUpdate": Object {
            "callback": null,
            "eventTime": 0,
            "lane": 1,
            "next": null,
            "payload": Object {
              "element": <ForwardRef>
                <transform />
              </ForwardRef>,
            },
            "tag": 0,
          },
          "shared": Object {
            "pending": null,
          },
        },
      },
      "child": [Circular],
      "childLanes": 0,
      "dependencies": null,
      "elementType": null,
      "firstEffect": [Circular],
      "flags": 256,
      "index": 0,
      "key": null,
      "lanes": 0,
      "lastEffect": [Circular],
      "memoizedProps": null,
      "memoizedState": Object {
        "element": <ForwardRef>
          <transform />
        </ForwardRef>,
      },
      "mode": 0,
      "nextEffect": null,
      "pendingProps": null,
      "ref": null,
      "return": null,
      "selfBaseDuration": 0,
      "sibling": null,
      "stateNode": FiberRootNode {
        "_debugRootType": "createLegacyRoot()",
        "callbackNode": null,
        "callbackPriority": 15,
        "containerInfo": Object {
          "children": Array [],
          "createNodeMock": [Function],
          "tag": "CONTAINER",
        },
        "context": Object {},
        "current": [Circular],
        "entangledLanes": 0,
        "entanglements": Array [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        "eventTimes": Array [
          -1,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        "expirationTimes": Array [
          -1,
          -1,
          -1,
          -1,
          -1,
          -1,
          -1,
          -1,
          -1,
          -1,
          -1,
          -1,
          -1,
          -1,
          -1,
          -1,
          -1,
          -1,
          -1,
          -1,
          -1,
          -1,
          -1,
          -1,
          -1,
          -1,
          -1,
          -1,
          -1,
          -1,
          -1,
        ],
        "expiredLanes": 0,
        "finishedLanes": 0,
        "finishedWork": null,
        "hydrate": false,
        "interactionThreadID": 1,
        "memoizedInteractions": Set {},
        "mutableReadLanes": 0,
        "pendingChildren": null,
        "pendingContext": null,
        "pendingInteractionMap": Map {},
        "pendingLanes": 0,
        "pingCache": null,
        "pingedLanes": 0,
        "suspendedLanes": 0,
        "tag": 0,
        "timeoutHandle": -1,
      },
      "tag": 3,
      "treeBaseDuration": 0,
      "type": null,
      "updateQueue": Object {
        "baseState": Object {
          "element": <ForwardRef>
            <transform />
          </ForwardRef>,
        },
        "effects": null,
        "firstBaseUpdate": null,
        "lastBaseUpdate": null,
        "shared": Object {
          "pending": null,
        },
      },
    },
    "selfBaseDuration": 0,
    "sibling": null,
    "stateNode": null,
    "tag": 11,
    "treeBaseDuration": 0,
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    },
    "updateQueue": Object {
      "lastEffect": Object {
        "create": [Function],
        "deps": Array [],
        "destroy": [Function],
        "next": Object {
          "create": [Function],
          "deps": Array [
            0,
            0,
            undefined,
            Object {},
            <transform />,
          ],
          "destroy": undefined,
          "next": [Circular],
          "tag": 3,
        },
        "tag": 5,
      },
    },
  },
}
`;
